### ISSUE TITLE
Refreshing state with many auth0_connection_client resources takes a long time

### ISSUE THREAD

**Issue by kherock** [auth0/terraform-provider-auth0#549]
### Checklist

- [X] I agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).

### Describe the problem you'd like to have solved

Since the removal of the `enabled_clients` property from the `auth0_connection` resource, I've migrated my configurations to use the `auth0_connection_client` resource. This migration is relatively painless with `for_each`:

```hcl
resource "auth0_connection_client" "google" {
  for_each = toset(local.enabled_clients)

  connection_id = auth0_connection.google.id
  client_id = each.value
}
```

However, it seems like instead of fetching the list of enabled connections once, Terraform is now spending time fetching the connection status individually for each client + connection pair. I have about 90 of these connections currently, and if I forget to pass `-refresh=false`, I can spend upwards of 5 minutes waiting for all of the resources to refresh!

### Describe the ideal solution

I don't know Terraform providers that well, but I hope that it would be possible to use a locally cached copy of enabled clients on a connection when resolving the status of an individual `auth0_connection_client` resource.

**Comment by raigow** [comment]
I think this would solve my issue with #546 , and wouldn't be too painful to switch to a new resource type. I assume that "we'd also have a way to raise awareness of when new clients are added to the connection and are not yet tracked through terraform" means that terraform would detect that it isn't managed and will want to destroy it by disabling. Is that accurate?

**Comment by sergiught** [comment]
@raigow That is correct, by switching over from `auth0_connection_client` (tracking 1:1 associations) to `auth0_connection_clients` (tracking 1:many associations), terraform will be able to show you an accurate drift between what clients you have enabled on the connection through your config and what actually exists within Auth0. 

If the drift will show that you have additional clients not tracked on the enabled_clients property of the connection and you proceed with a terraform apply regardless, it will disable the additional ones on Auth0 but not in your terraform config. 

The limitation here is that using both auth0_connection_client and auth0_connection_clients to manage enabled clients on a connection will conflict, so you'll have to pick 1 method. We'll continue to offer both types of association tho. 

Let me know if you have any other questions.

**Comment by raigow** [comment]
Sounds like it solves our problem, thanks! 

You mentioned above that it may be a heavy lift. What do you think a timeframe would be for something like this?

**Comment by sergiught** [comment]
If we go forward with that option a possible ETA would be by the end of May, potentially sooner. I'll circle back with the team and update this issue as soon as I have more information.

To offer some additional context, we're ramping up resources to work on pushing this provider to v1 in the upcoming months.

**Comment by kherock** [comment]
> So my question to you, would 1) solve your issue? Could you give us more context on how you're managing your clients and connections? 🙏🏻 We'd really appreciate it so we can come up with the best solution for yourself and everyone else using this provider.

I'm actually not super interested in 1) because I have since refactored everything to use the modular approach as it is much more ergonomic to set up that way for my use. I'm actually not particularly interested in having Terraform fully manage the `enabled_clients` list, mainly because I want to allow other root Terraform modules to deploy clients and client connections to the same tenant.

I do have the same problem as #546, but I would like to have it solved by an `auth0_client_connections` resource that fully manages the list of connections enabled for a client. I believe that refreshing this is possible with a single request to the `/connections` search endpoint.

My suggestion is to create `auth0_connection_clients` as you suggested, but deprecate `auth0_connection_client` for `auth0_client_connections` that fixes #546. Ideally, I'd like to have both options available as they both cover valid cases:

* `auth0_connection_clients` - fully manages the list of enabled clients for an individual connection. If a connection has clients managed outside of Terraform, use `auth0_client_connections` instead
* `auth0_client_connections` - fully manages the list of enabled connections on an individual client

**Comment by sergiught** [comment]
That's super valuable feedback @kherock thanks! 

So if I'm understanding this right, you're proposing to invert the current 1:1 relation between connection and client and expand it so you can attach a client to as many connections as you want. 

So `auth0_connection_client` (1 connection to 1 client) would become `auth0_client_connections`  (1 client to many connections) and it could look something like:

```terraform
resource auth0_client_connections my_clients_connections {
    client_id = "theClientID"
    connection_ids = ["conn1", "conn2", ... ]
}
```

> **Warning** 1
> Please note that if we deprecate `auth0_connection_client` you'd have to eventually migrate all your config to use `auth0_client_connections` instead. However we'll make sure to just deprecate it but not remove it entirely. It will get removed only once we release v1. 

> **Warning** 2
> We would still need to perform 1 API request / connection to attach the client to each connection. 

Additionally also to add the `auth0_connection_clients` resource which would keep track of the association between 1 connection to many clients. Something like:

```terraform
resource auth0_connection_clients my_connections_clients {
    connection_id = "theConnID"
    client_ids = ["clientID1", "clientID2", ... ]
}
```

Please correct any of the above if they are wrong.

**Comment by kherock** [comment]
@sergiught that all looks right to me. The modularity is definitely worth the performance tradeoff. Just as an API suggestion, it might be better to design the API of `auth0_connection_clients` as

```hcl
resource auth0_connection_clients my_connections_clients {
    connection_id = "theConnID"
    enabled_clients = ["clientID1", "clientID2", ... ]
}
```

to better communicate that it's fully managing the `enabled_clients` property on the connection?

**Comment by sergiught** [comment]
Hey folks this is now available in the latest release https://github.com/auth0/terraform-provider-auth0/releases/tag/v0.47.0. 

Documentation: 

- https://registry.terraform.io/providers/auth0/auth0/latest/docs/resources/connection_clients

This new resource is able to manage the entire list of enabled clients on a connection and report changes. It will also do the changes in bulk in one API call instead of separate API calls using the 1:1 `auth0_connection_client` resource.

While inverting the relationship and having a,

```
resource auth0_client_connections my_clients_connections {
    client_id = "theClientID"
    connection_ids = ["conn1", "conn2", ... ]
}
```

would be nice to have, it would offer no benefit over the current `auth0_client_connection` resource, as every client would need to get added to every connection in a separate API call, so we decided to not include it at least for now.

Let us know if you encounter any issues with the new attachment resources! 

Appreciate everyones patience while we worked on adding the new feature.
